### DISCUSSION TITLE
Clip as markdown

### DISCUSSION THREAD

**Discussion by arunnbabu81** [buggyj/tiddlyclip-plugin#79] [Ideas]
Currently there is an option to clip from website as html. Is it a viable option to add clip from website as markdown. I found an addon called [Markdownload](https://addons.mozilla.org/en-US/firefox/addon/markdownload/) which download webpages in markdown format. I think many people will be interested in clipping webpages as markdown directly into tiddlywiki. 

You may shift this discussion to the [tiddlyclip addon repo](https://github.com/buggyj/tiddlyclip)

**Comment by buggyj** [comment]
Conversion would be done in the tiddlywiki not the addon. There is a TC macro to convert to TW2 that I wrote some years ago - `@htmlToTW2(@web)`. I never found it useful to covert - often the conversion is not good, and with the ckeditor I found it easier to add annotations to the html. 
Also I do not see how markdown would be more useful?

**Reply by arunnbabu81** [reply]
> Conversion would be done in the tiddlywiki not the addon

I was asking whether the add-on can have the ability to clip the webpage content in markdown format and then send it to wiki (Similar to the MarkDownload addon I shared)


> Also I do not see how markdown would be more useful?

Personally I haven't used markdown yet. But I have seen discussions in forums, that markdown is more future proof. Don't know how true is that.

> There is a TC macro to convert to TW2 that I wrote some years ago - `@htmlToTW2(@web)`. I never found it useful to covert - often the conversion is not good, and with the ckeditor I found it easier to add annotations to the html.

I have seen a macro written by Mark to convert html to wikitext in the google group. I have tried to convert some article which were clipped as html using tiddlyclip to wikitext - although the text part was converted ok, the images were couldn't be rendered after the conversion. So my experience with conversion is also not that great. 

Another thing I noticed with clipped html tiddler were that - if I open them in ckeditor, there was not much lag even if the html tiddler in large. But if I tried to open them in the default TW editor, there was some noticeable lag . So I was confused whether to clip large articles as html. That's why I thought of markdown format. How is your experience with  clipping of large articles as html.



> and with the ckeditor I found it easier to add annotations to the html.

Mark is [developing is an annotator ](https://talk.tiddlywiki.org/t/has-someone-already-made-a-dynanote-annotater/2349) based on the [official dyna-annotate plug in](https://tiddlywiki.com/prerelease/#%24%3A%2Fplugins%2Ftiddlywiki%2Fdynannotate). May be that can improve the annotation capabilities in TW.

**Comment by buggyj** [comment]
markdown is used in a different context to tw5 markup. tw5 is really a software development language, a bit like react. Markdown is used where some formatting of text is required but html must be avoided for security reasons. 

I found annotating (in general) not to be very useful - it takes a lot of effort and I did not find the annotations very useful when re-reading (a part from when I am learning foreign languages). However with the annotations I make with ckeditor I can extract them from the text using tiddlyclip. 

How  big is a large clip? Ckeditor4 was the best editor for adding to webpages it was developed by a team of software engineers, i.e a lot of time and effort was put into making to work well so I am not surprised that it works better that the tw editor with large texts. 

I tend to take multiple clips from different webpages on the subject I am interested in and then write some separate notes (sometimes) based upon the clips (which can be video/audio as well as text.). I do this occasionally for a client. I also create 'paper trails' of the websites I visit as part of my research.

**Reply by arunnbabu81** [reply]
> However with the annotations I make with ckeditor I can extract them from the text using tiddlyclip.

How to do that?

> I tend to take multiple clips from different webpages on the subject I am interested in and then write some separate notes (sometimes) based upon the clips (which can be video/audio as well as text.). I do this occasionally for a client. I also create 'paper trails' of the websites I visit as part of my research.

Can you share a sample of how you take notes in this way

> How big is a large clip?

This is an example of the large clip I took

https://pubs.rsna.org/doi/10.1148/rg.2020190085

I used `select `feature to select the full text. I was experimenting with clipping of the entire article like this. Is that a bad strategy. Should I clip only the  relevant text only

**Comment by Marxsal** [comment]
I use an extension called "copy cat" to copy selected text to markdown. (there are other extensions like this). Then I use tiddlyclip and the clipboard option to clip the contents of the clipboard. I have this set up to create the markdown field type. So it's only one extra step over directly clipping in markdown. My rule looks like this:

```
\rules only table wikilink prettylink

|!Title|!Body|!Tags|!Fields modifiers|!field initials|!Modes|
|((*@pageTitle*))|((*@pageRef*))\n\n((*@clip*))\n\n|Articles 2022||{"$url":"((*@pageRef*))"},{"$caption":"((*@pageTitle*))"},{"$when":"((*@dateTimeShort*))"},{"$type":"text/x-markdown"}|append add|
```
My feeling is that most HTML that you clip is "noisy", taking  up to 3 times or more the space that the same text in markdown would take. 

I do wish there was official support for WYSIWYG in TW. I really thought that TW5 was going to have a WYSIWYG editor available, at least in the plugin library.

**Comment by buggyj** [comment]
@Marxsal   It is good to see people making their own methods of working with tiddlyclip(TC). I have tried to make TC as configurable as possible while still having some basic functionality out-of-the-box. tiddlyclip is not opinionated about the clips, allowing the user to decide how to created them. Personally I do not have any need for markdown conversion at this time, but when I do I will definitely consider you method.   

A Webpage's html source-code can be extremely complex, so when we clip a part of the page(as html) it is not surprising that we get a lot of html markup (often broken as it is incomplete). I find that ckeditor is very good at handle these fragments of html, and it enables me to hightlight sections easily.

The first version of the ckeditor adpater plugin was release 8 years ago - I look from time to time to see if a better js html editor  is available but I have not found a more suitable one.

**Comment by buggyj** [comment]
@arunnbabu81 - I try to use tiddlyclip to  condense information I find interesting on the web, so I try to make the smallest clips, however sometime this is a large amount.

I use tiddlyclip with a number of other plugins that I have written - my aim is to keep clips separate from my own writings, and so I put clips inside plugin tiddlers which I call 'closedlists'.

https://github.com/buggyj/tw5-other/tree/closedlist/xplugins/bj/closedlist

**Reply by arunnbabu81** [reply]
> @arunnbabu81 - I try to use tiddlyclip to condense information I find interesting on the web, so I try to make the smallest clips, however sometime this is a large amount.
> 
> I use tiddlyclip with a number of other plugins that I have written - my aim is to keep clips separate from my own writings, and so I put clips inside plugin tiddlers which I call 'closedlists'.
> 
> https://github.com/buggyj/tw5-other/tree/closedlist/xplugins/bj/closedlist

How to try this plug in

**Reply by Marxsal** [reply]
> I use tiddlyclip with a number of other plugins that I have written - my aim is to keep clips separate from my own writings, and so I put clips inside plugin tiddlers which I call 'closedlists'.

I have 'closedlists' and 'typestemplate' installed. But I don't have a clue how to  use or configure them. Nothing special seems to happen when I look at plugins. In addition, the readme for cllosedlists says "For edit, Needs ckeditor with subtiddler extension and multtid tools". But I don't know what the subtiddler extension is nor how to install it.

Thanks!

**Reply by buggyj** [reply]
you need these plugins and tiddler:
[TesT1_Clips.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/8193555/TesT1_Clips.json.zip)
These Only work up to 5.1.21 - I am updating to lastest TW release, but it will take me a couple of hours at least.

With this system you need to dock to the tiddlywiki, and then dock to the tiddler test1/clip  (by pressing the button on the tiddler) and then the clips will appear in the test1/clip tiddler.

**Reply by buggyj** [reply]
demo here - https://closedlist.tiddlyhost.com/

**Reply by Marxsal** [reply]
It looks interesting, but I'm not completely sure how to use it.

I set a snap area and tried "Snappy". It created an entry, but no image. 

I guess the tags option limits which items you can see in a closed list? But unfortunately, you can't set the tag on an item 

When I edited a closedlist and then clicked on an item, I got the message "loading ckeditor" -- but nothing came up. Which means you can't change things like the tags.

The search will tell you which closed list has your search term, but you don't know which item in the list. If you had a long list that could be a problem.

Thanks!

**Reply by buggyj** [reply]
My apologies for zero documentation. 

If you download the tiddlywiki locally you have to set a path to ckeditor library. If you have not installed ckeditor library locally you can use the online version by modifying this tiddler like so:

![image](https://user-images.githubusercontent.com/2855795/157201966-acf500bc-979b-43c7-b7e8-6957709b96b9.png)

**Reply by buggyj** [reply]
Sometime the ckeditor library online at http://cdnjs.clooudflare.com is not available and the message "loading ckeditor" appears.

**Reply by buggyj** [reply]
The closelist default snap rule saves the img external using the save media plugin as mentioned here:
https://github.com/buggyj/tiddlyclip-plugin/discussions/51

All the closedlist 'tiddlyclipconfig' starts in this tid:

`$:/bj/closelistclipconfig`

**Reply by buggyj** [reply]
The tags option are appear  at the bottom of the tiddlyclip dropdown in the same manner as discussed here:

https://github.com/buggyj/tiddlyclip-plugin/discussions/59

this allows clips to be tagged.

**Reply by buggyj** [reply]
I will add add a search function at some point.

I have hundreds of closedlist tiddlers - each on a particular narrow subject. Each one has many clips inside that serve as an extended reference on a subject. Sometimes I take many clips in a day. Closedlist stop my tiddlywikis becoming disorganised with potentially thousands of clips.

**Reply by buggyj** [reply]
I also have another plugin that works with tiddlyclip - 'Tickets' that I use more often than closedlist, it looks like this

![image](https://user-images.githubusercontent.com/2855795/157215082-b8f591a6-d0ae-47bc-966a-84a566f83af2.png)

**Reply by Marxsal** [reply]
> If you download the tiddlywiki locally you have to set a path to ckeditor library. If you have not installed ckeditor library locally you can use the online version by modifying this tiddler like so

The confusing thing is, includlib is already populated, but in script tags. But that vsn doesn't seem to work.

**Reply by Marxsal** [reply]
> The tags option are appear at the bottom of the tiddlyclip dropdown in the same manner as discussed here:
> 
> #59
> 
> this allows clips to be tagged.

What if you forget to apply those tags?  Or think of an additional one later?

On the tags, you have checkbox "extra tags". What does that do?

**Reply by arunnbabu81** [reply]
> I also have another plugin that works with tiddlyclip - 'Tickets' that I use more often than closedlist, it looks like this
> 
> ![image](https://user-images.githubusercontent.com/2855795/157215082-b8f591a6-d0ae-47bc-966a-84a566f83af2.png)

Could you share the tickets plug-in also.  How is it used.

On a sidenote , what is [searchtids ](https://github.com/buggyj/searchtids)used for

**Reply by buggyj** [reply]
I need to update the plugin to work with the lastest TW releases.  As you can see there are a lot of options. When I have updated it I will put a demo on tiddlyhost.com with some notes.

searchtids is a electron based desktop app that I am writing - it allows me to search all my tiddlywikis I have on my computer for tiddlers base on content, tags and title.

**Reply by arunnbabu81** [reply]
> I need to update the plugin to work with the lastest TW releases. As you can see there are a lot of options. When I have updated it I will put a demo on tiddlyhost.com with some notes.
> 


Will the closedlist plug in work with T1 5.2.1 presently. I tried it in my wiki based on 5.2.1, but i could not find the add-on context menu as seen in the demo site (as shown in the image below).

![image](https://user-images.githubusercontent.com/67494083/157299414-264aaa57-2a27-4092-bbe8-1904a9ecb9b4.png)

> searchtids is a electron based desktop app that I am writing - it allows me to search all my tiddlywikis I have on my computer for tiddlers base on content, tags and title.
>
Eager to test it out once you finish the development of the searchtids

**Reply by buggyj** [reply]
you need all the plugins that are on the demo page. - BUT this is not a release version there are some issues (eg navigation in story not working), and I have to decide whether to support the new template cascade (presently it depends on the types template plugin)

**Reply by arunnbabu81** [reply]
> searchtids is a electron based desktop app that I am writing - it allows me to search all my tiddlywikis I have on my computer for tiddlers base on content, tags and title.

@buggyj Is this available for testing?

Anything similar possible for mobile?

**Reply by buggyj** [reply]
there is a problem with a library that I have used - I need to find a replacement but at present I am prioritizing tiddlyclip.

 I do not use tiddlywiki on my mobile, so I have not investigated this for the mobile.

**Comment by buggyj** [comment]
### New tags can be added via editing the closedlist tiddler, eg here 'newtag' has just been added - it needs to appear in both places indicated. (notice the comma separator on the lower field)

![image](https://user-images.githubusercontent.com/2855795/157292703-a7c1b410-5ff2-43af-9d64-fe6abe282ad9.png)

### The new tags then appears after saving:

![image](https://user-images.githubusercontent.com/2855795/157292939-53849ee4-8434-4417-9b68-bd08c0d27218.png)

### clicking on a tag pill shows the clips that have that tag, the tag is remove by click on the small cross in a circle on the right side. tags are added by dragging a tag pill and dropping it on the large cross on the left side.

![image](https://user-images.githubusercontent.com/2855795/157293471-8f4e1edd-9068-47f8-9c21-c065d60f25a3.png)

**Reply by Marxsal** [reply]
When I try to use "snappy", something briefly appears in the side bar making me think download is happening. But when I navigate to the docked page there is an alert that says "This page says" -- but no content. No file appears in the "media" sub directory.

Thanks!

**Reply by buggyj** [reply]
I use savemedia together with savetiddlers so myTWs are in a sub-directory of the download directory.

**Reply by Marxsal** [reply]
For these experiments I have file:

`file:///home/mark/temp/tiddlywiki/bj/closed-list-demo.html`

and download directory set to

`/home/mark/temp/tiddlywiki/bj/`

Does savemedia need savetiddlers to work? 

This is chromium. There's a note in your readme, but I didn't understand what the implications were. I'm pretty sure I got savemedia to work a few weeks ago. 

Thanks!

**Reply by buggyj** [reply]
savemedia is not dependent on savetiddlers. you can observe the savemedia work - go to chrome://extensions and chick on the 'background page' button and then select the console tab - then do a snap and see what appears

![image](https://user-images.githubusercontent.com/2855795/157332501-811a0aa8-adb0-4fcc-8795-37b60c3f2001.png)

**Reply by Marxsal** [reply]
Ok, so the media saver wants the download directory to be **2** levels above the media directory. Or at least that's what it took for the media to save. This was different from the tiddlersaver, which was fine with having the save directory be the same as the download directory.

**Comment by Marxsal** [comment]
There's something weird going on with the tagging. I created a "test" tag and then dragged it on to an article that was untagged. Immediately it disappeared. Now I can see the item only if I click on "all tags". But then I can't tag it because there's no tagging dialog. If I click on the "test" tag, the item I tagged doesn't show up -- the list is empty!

I think the lack of transparency re the tagging would be very hard to deal with long term.

With regular plugins I can access the shadow tiddler and examine it. But with the closed list tiddler I'm kind of barred from any sort of tweaking or fixing.

**Reply by buggyj** [reply]
I guess it is pretty confusing! The interface is a template it can be improved... I have made some changes to make the tags visible and added a delete-tag drop icon (cross in a circle), I will upload it when later today.

![image](https://user-images.githubusercontent.com/2855795/157467471-3de27381-d6a6-4200-8b37-9369ff1bc4a2.png)

**Reply by buggyj** [reply]
@arunnbabu81 @Marxsal  I have put a demo here:

https://closedlistdemo.tiddlyhost.com/
